lls/main.c

```diff
@@ -358,6 +359,151 @@ match_nd_router_sol_or_adv(struct rte_mbuf *pkt, struct gatekeeper_if *iface)
 	return 0;
 }
 
+unsigned short
+icmp_cksum(void *buf, unsigned int size)
+{
+	unsigned short *buffer = buf;
+	unsigned long cksum = 0;
+
+	while(size > 1) {
+		cksum += *buffer++;
+		size -= sizeof(*buffer);
+	}
+
+	if(size)
+		cksum += *(unsigned char*)buffer;
+
+	cksum = (cksum >> 16) + (cksum & 0xffff);
+	cksum += (cksum >> 16);
+
+	return (unsigned short)(~cksum);
+}
+
+static void
+xmit_icmp_reply(struct gatekeeper_if *iface, struct rte_mbuf *pkt)
+{
+	struct ether_addr eth_addr_tmp;
+	struct ether_hdr *icmp_eth = rte_pktmbuf_mtod(pkt, struct ether_hdr *);
+	uint32_t ip_addr_tmp;
+	struct ipv4_hdr *icmp_ipv4;
+	struct icmp_hdr *icmph;
+
+	ether_addr_copy(&icmp_eth->s_addr, &eth_addr_tmp);
+	ether_addr_copy(&icmp_eth->d_addr, &icmp_eth->s_addr);
+	ether_addr_copy(&eth_addr_tmp, &icmp_eth->d_addr);
+	if (iface->vlan_insert)
+		fill_vlan_hdr(icmp_eth, iface->vlan_tag_be, ETHER_TYPE_IPv4);
+
+	/*
+	 * Set-up IPv4 header.
+	 *
+	 * According to RFC 792 page 13: to form an echo reply
+	 * message, the source and destination addresses are simply reversed,
+	 * the type code changed to 0, and the checksum recomputed.
+	 */
+	icmp_ipv4 = (struct ipv4_hdr *)pkt_out_skip_l2(iface, icmp_eth);
+	icmp_ipv4->time_to_live = IP_DEFTTL;
+	ip_addr_tmp = icmp_ipv4->src_addr;
+	icmp_ipv4->src_addr = icmp_ipv4->dst_addr;
+	icmp_ipv4->dst_addr = ip_addr_tmp;
+
+	/*
+	 * The IP header checksum filed must be set to 0
+	 * in order to offload the checksum calculation.
+	 */
+	icmp_ipv4->hdr_checksum = 0;
+	pkt->l2_len = iface->l2_len_out;
+	pkt->l3_len = sizeof(struct ipv4_hdr);
```

Comment:
What if there is an IPv4 extension header in the packet? We either take care of the whole IPv4 header (e.g. we are ignoring the total length field), or properly jump the IPv4 header.

This comment also applies to the line `icmph = (struct icmp_hdr *)&icmp_ipv4[1];` below.